applications/OptimizationApplication/python_scripts/utilities/response_function_implementor.py

```diff
@@ -0,0 +1,194 @@
+
+class ConstraintResponseFunctionImplementor(ResponseFunctionImplementor):
+    def __init__(self, parameters: Kratos.Parameters, optimization_info: OptimizationInfo):
+        default_parameters = Kratos.Parameters("""{
+            "response_name": "",
+            "type"         : "",
+            "scaling"      : 1.0,
+            "ref_type"     : "",
+            "ref_value"    : 0.0
+        }""")
+        parameters.ValidateAndAssignDefaults(default_parameters)
+
+        self.__scaling = parameters["scaling"].GetDouble()
+
+        self.__ref_type = parameters["ref_type"].GetString()
+        if self.__ref_type == "initial_value":
+            self.__reference_value = None
+        elif self.__ref_type == "specified_value":
+            self.__reference_value = parameters["ref_value"].GetDouble()
+        else:
+            raise RuntimeError(f"Provided \"reference_type\" = {self.__ref_type} is not supported for constraint response functions. Followings are supported options: \n\tinitial_value\n\tspecified_value")
+
+        self.__constraint_type = parameters["type"].GetString()
+        if self.__constraint_type in ["<", "<=", "="]:
+            self.__standardization_value = 1.0
+        elif self.__constraint_type in [">", ">="]:
+            self.__standardization_value = -1.0
+        else:
+            raise RuntimeError(f"Provided \"type\" = {self.__constraint_type} is not supported in constraint response functions. Followings are supported options: \n\t=\n\t<=\n\t<\n\t>=\n\t>")
+
+        super().__init__(parameters["response_name"].GetString(), optimization_info)
+
+    def GetReferenceValue(self):
+        if self.__reference_value is None:
+            self.__reference_value = self.GetValue()
+
+        return self.__reference_value
+
+    def GetStandardizedValue(self, solution_step_index: int = 0) -> float:
+        return self.__standardization_value * (self.GetValue(solution_step_index) * self.__scaling - self.GetReferenceValue())
+
+    def GetStandardizedSensitivity(self, sensitivity_variable, sensitivity_data_container: ContainerVariableDataHolderUnion, solution_step_index: int = 0):
+        self.GetSensitivity(sensitivity_variable, sensitivity_data_container, solution_step_index)
+        sensitivity_data_container *= (self.__standardization_value * self.__scaling)
+
+    def IsActive(self, solution_step_index: int = 0) -> bool:
+        return (self.__constraint_type == "=") or self.GetStandardizedValue(solution_step_index) >= 0.0
+
+    def GetResponseType(self) -> str:
+        return self.__constraint_type
+
+    def GetResponseInfo(self) -> str:
+        if self._optimization_info["step"] > 1:
+            relative_change = (self.GetValue() / self.GetValue(1) - 1.0)
+        else:
+            relative_change = 0.0
+        absolute_change = (self.GetValue() / self._GetInitialResponseValue() - 1.0)
+
+        if self.IsActive():
+            violation = self.GetStandardizedValue() / self.GetReferenceValue()
```

Comment:
You need to check if self.GetReferenceValue() is not zero.

Reply:
done :)